koku/koku/middleware.py

```diff
@@ -161,38 +184,87 @@ def process_request(self, request):
                 if username not in USER_CACHE:
                     USER_CACHE[username] = request.user
                     LOG.debug(f"User added to cache: {username}")
-                if not request.user.admin and request.user.access is None:
-                    LOG.warning("User %s is does not have permissions for Cost Management.", username)
-                    # For /user-access we do not want to raise the exception since the API will
-                    # return a false boolean response that the platfrom frontend code is expecting.
-                    if request.path != reverse("user-access"):
-                        raise PermissionDenied()
+                self._check_user_has_access(request)
+
             else:
                 return HttpResponseUnauthorizedRequest()
+
         try:
-            super().process_request(request)
+            tenant = self._get_or_create_tenant(request)
+            # Inherited from superclass. Set the tenant for the request
+            request.tenant = tenant
+            connection.set_tenant(request.tenant)
+
         except OperationalError as err:
             LOG.error("Request resulted in OperationalError: %s", err)
             DB_CONNECTION_ERRORS_COUNTER.inc()
             return HttpResponseFailedDependency({"source": "Database", "exception": err})
 
-    def get_tenant(self, model, hostname, request):
-        """Override the tenant selection logic."""
-        schema_name = "public"
+    def _check_user_has_access(self, request):
+        """Check if the user has access to Cost Management.
+
+        Args:
+            request (HttpRequest): The incoming request object.
+
+        Raises:
+            PermissionDenied: If the user does not have permissions for Cost Management.
+
+        """
+
+        username = request.user.username
+        if not request.user.admin and request.user.access is None:
+            LOG.warning("User %s does not have permissions for Cost Management.", username)
+            # For /user-access we do not want to raise the exception since the API will
+            # return a false boolean response that the platfrom frontend code is expecting.
+            if request.path != reverse("user-access"):
+                raise PermissionDenied()
+
+    def _get_or_create_tenant(self, request):
+        """Get or create tenant based on the user's schema.
+
+        Args:
+            request(HttpRquest): The incoming request object.
+
+        Returns:
+            Tenant: The tenant object.
+
+        """
+
+        if tenant := self._get_tenant_from_tenant_cache(request):
+            return tenant
+
+        tenant_username = request.user.username
+        schema_name = request.user.customer.schema_name if not is_no_auth(request) else "public"
+        if tenant := self._get_tenant_from_db(tenant_username, schema_name):
+            return tenant
+
+        return self._create_tenant()
```

Comment:
If `_get_tenant_from_db()` is changed to raise rather than returning `None`:
```suggestion
        try:
            return self._get_tenant_from_db(tenant_username, schema_name)
        except Tenant.DoesNotExist:
            LOG.info("No tenant found. Creating")  # Just an example message.
            return self._create_tenant()
```

Reply:
Thanks for this suggestion as well, reduces unnecessary variable definitions.